Switching to Java 17

jacekbilski opened the pull request:
Preparations before upgrading Spring Boot to version 3.0.
I also cleaned up some warnings here and there. Not all of them, that would require getting rid of `org.springframework.security.oauth.boot:spring-security-oauth2-autoconfigure`

jacekbilski committed:
Removing most of Gradle deprecations/warnings.

jacekbilski committed:
Fixed some Java-related deprecations.

jacekbilski committed:
Switching tests to use JUnit5 APIs. Oleaster stays as-is, we're using the newest version anyway.

jacekbilski committed:
Switching to Java 17.

erikpetzold commented on src/main/java/org/redischool/attendance/details/CourseHelper.java:
```diff
@@ -30,7 +30,7 @@ public String closestCourseDate(Date targetDate, CourseDetails courseDetails) {
         List<Object> datesMap = new ArrayList<>();
         for (int i = 0; i < courseDetails.getJavaDates().size(); i++) {
-            Map dateMap = new HashMap<String, List<String>>();
+            Map<String, Object> dateMap = new HashMap<>();
```
why `Map<String, Object>` and not `Map<String, List<String>>` like it was on right side before?

jacekbilski replied:
Fixed in 5d4cae9. It's actually not `Map<String, List<String>>` but `Map<String, String>`. You can see that in lines 41 and 42. In declaration in line 33 the type on the right hand side doesn't matter anyway because of type erasure, therefore it could have been wrong the whole time.

jacekbilski committed:
Fixing types in `CourseHelper.getFormattedDatesMap`.
Value of `datesMap` in `CourseDetailsControllerTests.testRenderCourseDetails` makes no difference to the test, therefore using empty. The test is actually not verifying much.

Rappahannock commented on src/test/java/org/redischool/attendance/FeatureTestForSecurity.java:
```diff
@@ -24,7 +24,7 @@
         }
 )
-@RunWith(SpringJUnit4ClassRunner.class)
+@ExtendWith(SpringExtension.class)
```
It is already included in the @SpringBootTest, can be deleted

jacekbilski replied:
Fixed.

Rappahannock commented on src/test/java/org/redischool/attendance/FeatureTests.java:
```diff
@@ -25,11 +25,10 @@
 
 @SpringBootTest(webEnvironment = RANDOM_PORT)
-@RunWith(SpringJUnit4ClassRunner.class)
+@ExtendWith(SpringExtension.class)
```
It is already included in the @SpringBootTest, can be deleted

jacekbilski replied:
Fixed.

jacekbilski committed:
Documenting that this project requires Java 17.

Rappahannock commented:
Would be nice to update README to mention Java 17 instead of Java 11.

Rappahannock commented:
> Would be nice to update README to mention Java 17 instead of Java 11.

I see I was late to add this comment and it has already been added :-)

jacekbilski committed:
TIL, `@SpringBootTest` already includes `@ExtendWith(SpringExtension.class)`.
